### PR TITLE
Use If-None-Match in LakeFSLinker

### LINKED ISSUE #7488: lakeFSFS: Use If-None-Match in LakeFSLinker
Allow using If-None-Match as overwrite control in LakeFSLinker for greater concurrency control

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #7488

### Background

Allow creating Data/Output/InputStream with conditional overwrite on the last stage of creating the lakeFS entry by supporting the If-None-Match in the LakeFSLinker

### Testing Details

Should be covered by the Hadoop contractual testing

### Breaking Change?

No

### CODE REVIEW

File: clients/hadoopfs/src/main/java/io/lakefs/storage/SimpleStorageAccessStrategy.java [hunk 1]
```diff
@@ -37,7 +37,13 @@ public SimpleStorageAccessStrategy(LakeFSFileSystem lakeFSFileSystem, LakeFSClie
 
     @Override
     public FSDataOutputStream createDataOutputStream(ObjectLocation objectLocation,
-            CreateOutputStreamParams params) throws ApiException, IOException {
+                                                     CreateOutputStreamParams params) throws ApiException, IOException {
+        return createDataOutputStream(objectLocation, params, true);
+    }
```
Comment (arielshaqed) [inline clients/hadoopfs/src/main/java/io/lakefs/storage/SimpleStorageAccessStrategy.java:42]: This should now become an [interface default method](https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html).  I think those are already available in our Java versions.

**Review by arielshaqed — COMMENTED** [review]
Thanks!

Aside from the comment, I'm not sure how catalogs will be able to consume this method.  It's not accessible via any FileSystem method.

**Comment by N-o-Z** [comment]
> Thanks!
> 
> Aside from the comment, I'm not sure how catalogs will be able to consume this method. It's not accessible via any FileSystem method.

Catalogs consume this method via the TableMetadataParser.write and the FIleIO implementation. as part of the write method we invoke either a create method or a createOrOverwrite method which produces a FSDataOutputStream with the overwrite param set accordingly

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Looking for references to back me up on "overwrite".

File: clients/hadoopfs/src/main/java/io/lakefs/LakeFSLinker.java [hunk 2]
```diff
@@ -13,23 +13,28 @@ public class LakeFSLinker {
...
                 new StagingMetadata().checksum(eTag).sizeBytes(byteSize).staging(stagingLocation);
         try {
-            staging.linkPhysicalAddress(objectLoc.getRepository(), objectLoc.getRef(),
-                    objectLoc.getPath(), stagingMetadata)
-                .execute();
+            StagingApi.APIlinkPhysicalAddressRequest request = 
+                    staging.linkPhysicalAddress(objectLoc.getRepository(), objectLoc.getRef(), objectLoc.getPath(), stagingMetadata);
+            if (!overwrite) {
+                request.ifNoneMatch("*");
+            }
```
[the remark this answers is not in the document]
Comment (arielshaqed) [reply clients/hadoopfs/src/main/java/io/lakefs/LakeFSLinker.java:36]: Turns out that I was right!  I don't understand this _at **all**_.  This _literally_ implements the semantics that I asked for.

**Review by arielshaqed — APPROVED** [review]
Thanks!

Just the interface default method thing, and let's go publish!

**Pull request merged**